retrieve_assertion.go

```diff
@@ -66,17 +71,28 @@ func (sp *SAMLServiceProvider) RetrieveAssertionInfo(encodedResponse string) (*A
 		return nil, ErrMissingElement{Tag: AttributeStatementTag}
 	}
 
-	info := make(map[string]string)
-	for _, child := range attributeStatement.ChildElements() {
-		nameAttr := child.SelectAttr(NameAttr)
-		attributeValue := child.FindElement(childPath(child.Space, AttributeValueTag))
-		if attributeValue == nil {
-			return nil, ErrMissingElement{Tag: AttributeValueTag}
-		}
-		info[nameAttr.Value] = attributeValue.Text()
+	doc := etree.NewDocument()
+	doc.SetRoot(attributeStatement)
+	bs, err := doc.WriteToBytes()
+
+	if err != nil {
+		return nil, err
+	}
+
+	var attrs struct {
+		List []Attribute `xml:"Attribute"`
+	}
+	err = xml.Unmarshal(bs, &attrs)
+	if err != nil {
+		return nil, err
+	}
```

Comment:
Yeah, it's mostly because of the simplicity of unmarshal and my familiarity with that code style. In general, I tend to significantly prefer encoding/xml for unmarshaling, since it handles a lot of that quite well, and is highly declarative and simple to read, since you're often just looking at struct tags.

For marshalling, though, I think etree has the edge (especially in our case needing to canonicalize). I'm not familiar enough with implementing xml.Marshaler to do something that complex with any confidence.

I've actually been thinking about trying to contribute some code to migrate more of the code to this style, since the codebase is split a bit stylistically since the decryption code I integrated uses that method almost exclusively.

If you'd prefer consistency, though, I can attempt to rewrite this part using etree. Just let me know. :-)